src/consensus/vote_tracker.zig

```diff
@@ -0,0 +1,343 @@
+    pub const progressWithNewRootBank = purgeStaleState;
+
+    /// Purge any outdated slot data
+    pub fn purgeStaleState(
+        self: *VoteTracker,
+        allocator: std.mem.Allocator,
+        new_root_bank_slot: Slot,
+    ) void {
+        self.map_rwlock.lock();
+        defer self.map_rwlock.unlock();
+        const map = &self.map;
+
+        var start_idx: usize = 0;
+        outer: while (start_idx != map.count()) {
+            for (map.keys()[start_idx..], start_idx..) |slot, i| {
+                if (slot >= new_root_bank_slot) continue;
+                map.fetchSwapRemove(slot).?.value.deinit(allocator);
+                start_idx = i;
+                continue :outer;
+            }
+        }
+    }
+
+    /// NOTE: intended for tests
+    pub fn insertVote(
+        self: *VoteTracker,
+        allocator: std.mem.Allocator,
+        slot: Slot,
+        pubkey: Pubkey,
+    ) !void {
+        if (!builtin.is_test) @compileError(@src().fn_name ++ " can only be used in tests.");
+
+        self.map_rwlock.lock();
+        defer self.map_rwlock.unlock();
+        const map = &self.map;
+
+        const gop = try map.getOrPut(allocator, slot);
+        errdefer if (!gop.found_existing) std.debug.assert(map.pop().key == slot);
+        if (!gop.found_existing) gop.value_ptr.* = try RcRwSlotVoteTracker.create(allocator);
+        errdefer if (!gop.found_existing) gop.value_ptr.*.deinit(allocator);
+
+        const slot_vote_tracker, //
+        var slot_vote_tracker_lg //
+        = gop.value_ptr.*.tracker.writeWithLock();
+        defer slot_vote_tracker_lg.unlock();
+
+        map.lockPointers();
+        defer map.unlockPointers();
+
+        const voted = &slot_vote_tracker.voted;
+        errdefer if (!gop.found_existing) voted.deinit(allocator);
+
+        const voted_slot_updates = slot_vote_tracker.initAndOrGetUpdates();
+        errdefer if (!gop.found_existing) voted_slot_updates.deinit(allocator);
+
+        if (!voted.contains(pubkey)) try voted.ensureUnusedCapacity(allocator, 1);
+        try voted_slot_updates.ensureUnusedCapacity(allocator, 1);
+
+        voted.putAssumeCapacity(pubkey, true);
+        voted_slot_updates.appendAssumeCapacity(pubkey);
+    }
+};
+
+pub const SlotVoteTracker = struct {
+    /// Maps pubkeys that have voted for this slot
+    /// to whether or not we've seen the vote on gossip.
+    /// True if seen on gossip, false if only seen in replay.
+    voted: std.AutoArrayHashMapUnmanaged(Pubkey, bool),
+    optimistic_votes_tracker: std.AutoArrayHashMapUnmanaged(Hash, VoteStakeTracker),
+    voted_slot_updates: ?std.ArrayListUnmanaged(Pubkey),
+    gossip_only_stake: u64,
+
+    pub const EMPTY_ZEROES: SlotVoteTracker = .{
+        .voted = .{},
+        .optimistic_votes_tracker = .{},
+        .voted_slot_updates = null,
+        .gossip_only_stake = 0,
+    };
+
+    pub fn deinit(self: SlotVoteTracker, allocator: std.mem.Allocator) void {
+        var copy = self;
```

Comment:
Is this meant to work based on the assumption that `SlotVoteTracker` would be passed in as a shallow copy? and hence the fields can still be reached and de-allocated?

Reply:
I was actually asking about how self is passed in. ie `deinit(self: SlotVoteTracker..`? 

Was wondering about having `pub fn deinit(self: *SlotVoteTracker` instead? and why would that be a bad/good approach.

Reply:
Passing it in mutably would mean its mutability is viral. I prefer mutability not be required to deinitialize something is my point.

Reply:
The options available I believe are:

- Passing it as a mutable ref
- Passing it as an immutable ref
- Passing it as an immutable copy

And in your approach, you are using the last approach. My question was about how that works (and if that is safe) since it is being deinitialize via a copy.

Reply:
Yeah, it's perfectly safe. The pointers passed to `allocator.free` under the hood are still the same pointers, the only reason those methods take in a mutable reference to the data structures is to do `self.* = undefined`, which I view as pointless, and is the subject of debate as to whether that should even be the convention.
See `std.ArrayList(T).deinit`, which in fact doesn't do this, representing the current split in the debate.